File: java-sdk/src/main/java/com/framework/exception/ErrorCode.java
package com.framework.exception;

import java.util.HashMap;
import java.util.Map;

/**
 * 统一错误码枚举
 * 
 * 定义框架所有错误码，并提供与 HTTP 状态码、gRPC 状态码、JSON-RPC 错误码之间的映射。
 */
public enum ErrorCode {
    // 客户端错误 (4xx)
    BAD_REQUEST(400, "Bad Request", true),
    UNAUTHORIZED(401, "Unauthorized", true),
    FORBIDDEN(403, "Forbidden", true),
    NOT_FOUND(404, "Not Found", true),
    TIMEOUT(408, "Timeout", true),
    
    // 服务端错误 (5xx)
    INTERNAL_ERROR(500, "Internal Error", false),
    NOT_IMPLEMENTED(501, "Not Implemented", false),
    SERVICE_UNAVAILABLE(503, "Service Unavailable", false),
    
    // 框架错误 (6xx)
    PROTOCOL_ERROR(600, "Protocol Error", false),
    SERIALIZATION_ERROR(601, "Serialization Error", false),
    ROUTING_ERROR(602, "Routing Error", false),
    CONNECTION_ERROR(603, "Connection Error", false);
    
    private final int code;
    private final String message;
    private final boolean clientError;

    // HTTP 状态码到 ErrorCode 的映射
    private static final Map<Integer, ErrorCode> HTTP_STATUS_MAP = new HashMap<>();
    // gRPC 状态码到 ErrorCode 的映射
    private static final Map<Integer, ErrorCode> GRPC_STATUS_MAP = new HashMap<>();
    // JSON-RPC 错误码到 ErrorCode 的映射
    private static final Map<Integer, ErrorCode> JSONRPC_CODE_MAP = new HashMap<>();
    // 框架错误码到 ErrorCode 的映射
    private static final Map<Integer, ErrorCode> CODE_MAP = new HashMap<>();
    
    static {
        for (ErrorCode ec : values()) {
            CODE_MAP.put(ec.code, ec);
        }
        
        // HTTP 状态码映射
        HTTP_STATUS_MAP.put(400, BAD_REQUEST);
        HTTP_STATUS_MAP.put(401, UNAUTHORIZED);
        HTTP_STATUS_MAP.put(403, FORBIDDEN);
        HTTP_STATUS_MAP.put(404, NOT_FOUND);
        HTTP_STATUS_MAP.put(408, TIMEOUT);
        HTTP_STATUS_MAP.put(500, INTERNAL_ERROR);
        HTTP_STATUS_MAP.put(501, NOT_IMPLEMENTED);
        HTTP_STATUS_MAP.put(503, SERVICE_UNAVAILABLE);
        
        // gRPC 状态码映射 (io.grpc.Status.Code ordinal values)
        GRPC_STATUS_MAP.put(0, null);  // OK - 无错误
        GRPC_STATUS_MAP.put(1, INTERNAL_ERROR);      // CANCELLED
        GRPC_STATUS_MAP.put(2, INTERNAL_ERROR);      // UNKNOWN
        GRPC_STATUS_MAP.put(3, BAD_REQUEST);          // INVALID_ARGUMENT
        GRPC_STATUS_MAP.put(4, TIMEOUT);              // DEADLINE_EXCEEDED
        GRPC_STATUS_MAP.put(5, NOT_FOUND);            // NOT_FOUND
        GRPC_STATUS_MAP.put(7, FORBIDDEN);            // PERMISSION_DENIED
        GRPC_STATUS_MAP.put(12, NOT_IMPLEMENTED);     // UNIMPLEMENTED
        GRPC_STATUS_MAP.put(13, INTERNAL_ERROR);      // INTERNAL
        GRPC_STATUS_MAP.put(14, SERVICE_UNAVAILABLE); // UNAVAILABLE
        GRPC_STATUS_MAP.put(16, UNAUTHORIZED);        // UNAUTHENTICATED
        
        // JSON-RPC 错误码映射
        JSONRPC_CODE_MAP.put(-32700, BAD_REQUEST);        // Parse error
        JSONRPC_CODE_MAP.put(-32600, BAD_REQUEST);        // Invalid Request
        JSONRPC_CODE_MAP.put(-32601, NOT_FOUND);          // Method not found
        JSONRPC_CODE_MAP.put(-32602, BAD_REQUEST);        // Invalid params
        JSONRPC_CODE_MAP.put(-32603, INTERNAL_ERROR);     // Internal error
    }
    
    ErrorCode(int code, String message, boolean clientError) {
        this.code = code;
        this.message = message;
        this.clientError = clientError;
    }
    
    public int getCode() {
        return code;
    }
    
    public String getMessage() {
        return message;
    }
    
    public boolean isClientError() {
        return clientError;
    }
    
    public boolean isServerError() {
        return code >= 500 && code < 600;
    }
    
    public boolean isFrameworkError() {
        return code >= 600;
    }
    
    /**
     * 是否为可重试的错误
     */
    public boolean isRetryable() {
        return this == TIMEOUT || this == SERVICE_UNAVAILABLE || this == CONNECTION_ERROR;
    }
    
    /**
     * 根据框架错误码查找 ErrorCode
     */
    public static ErrorCode fromCode(int code) {
        return CODE_MAP.getOrDefault(code, INTERNAL_ERROR);
    }
    
    /**
     * 从 HTTP 状态码映射到 ErrorCode
     */
    public static ErrorCode fromHttpStatus(int httpStatus) {
        ErrorCode mapped = HTTP_STATUS_MAP.get(httpStatus);
        if (mapped != null) {
            return mapped;
        }
        if (httpStatus >= 400 && httpStatus < 500) {
            return BAD_REQUEST;
        }
        if (httpStatus >= 500) {
            return INTERNAL_ERROR;
        }
        return INTERNAL_ERROR;
    }
    
    /**
     * 从 gRPC 状态码映射到 ErrorCode
     */
    public static ErrorCode fromGrpcStatus(int grpcStatusCode) {
        ErrorCode mapped = GRPC_STATUS_MAP.get(grpcStatusCode);
        return mapped != null ? mapped : INTERNAL_ERROR;
    }
    
    /**
     * 从 JSON-RPC 错误码映射到 ErrorCode
     */
    public static ErrorCode fromJsonRpcCode(int jsonRpcCode) {
        ErrorCode mapped = JSONRPC_CODE_MAP.get(jsonRpcCode);
        return mapped != null ? mapped : INTERNAL_ERROR;
    }
    
    /**
     * 将 ErrorCode 映射到 HTTP 状态码
     */
    public int toHttpStatus() {
        return switch (this) {
            case BAD_REQUEST -> 400;
            case UNAUTHORIZED -> 401;
            case FORBIDDEN -> 403;
            case NOT_FOUND -> 404;
            case TIMEOUT -> 408;
            case INTERNAL_ERROR -> 500;
            case NOT_IMPLEMENTED -> 501;
            case SERVICE_UNAVAILABLE -> 503;
            case PROTOCOL_ERROR -> 502;
            case SERIALIZATION_ERROR -> 400;
            case ROUTING_ERROR -> 502;
            case CONNECTION_ERROR -> 503;
        };
    }
    
    /**
     * 将 ErrorCode 映射到 JSON-RPC 错误码
     */
    public int toJsonRpcCode() {
        return switch (this) {
            case BAD_REQUEST -> -32600;
            case NOT_FOUND -> -32601;
            case INTERNAL_ERROR, TIMEOUT, SERVICE_UNAVAILABLE -> -32603;
            case SERIALIZATION_ERROR -> -32700;
            default -> -32603;
        };
    }
}

File: golang-sdk/errors/error_code.go
package errors

// ErrorCode 统一错误码类型
type ErrorCode int

const (
	// 客户端错误 (4xx)
	BadRequest ErrorCode = 400
	Unauthorized ErrorCode = 401
	Forbidden ErrorCode = 403
	NotFound ErrorCode = 404
	Timeout ErrorCode = 408

	// 服务端错误 (5xx)
	InternalError ErrorCode = 500
	NotImplemented ErrorCode = 501
	ServiceUnavailable ErrorCode = 503

	// 框架错误 (6xx)
	ProtocolError ErrorCode = 600
	SerializationError ErrorCode = 601
	RoutingError ErrorCode = 602
	ConnectionError ErrorCode = 603
)

// String 返回错误码的字符串表示
func (e ErrorCode) String() string {
	switch e {
	case BadRequest:
		return "Bad Request"
	case Unauthorized:
		return "Unauthorized"
	case Forbidden:
		return "Forbidden"
	case NotFound:
		return "Not Found"
	case Timeout:
		return "Timeout"
	case InternalError:
		return "Internal Error"
	case NotImplemented:
		return "Not Implemented"
	case ServiceUnavailable:
		return "Service Unavailable"
	case ProtocolError:
		return "Protocol Error"
	case SerializationError:
		return "Serialization Error"
	case RoutingError:
		return "Routing Error"
	case ConnectionError:
		return "Connection Error"
	default:
		return "Unknown Error"
	}
}

// Code 返回错误码的整数值
func (e ErrorCode) Code() int {
	return int(e)
}

// IsClientError 判断是否为客户端错误
func (e ErrorCode) IsClientError() bool {
	return e >= 400 && e < 500
}

// IsServerError 判断是否为服务端错误
func (e ErrorCode) IsServerError() bool {
	return e >= 500 && e < 600
}

// IsFrameworkError 判断是否为框架错误
func (e ErrorCode) IsFrameworkError() bool {
	return e >= 600
}

// IsRetryable 判断是否为可重试的错误
func (e ErrorCode) IsRetryable() bool {
	return e == Timeout || e == ServiceUnavailable || e == ConnectionError
}

// FromCode 根据错误码整数值获取 ErrorCode
func FromCode(code int) ErrorCode {
	switch code {
	case 400:
		return BadRequest
	case 401:
		return Unauthorized
	case 403:
		return Forbidden
	case 404:
		return NotFound
	case 408:
		return Timeout
	case 500:
		return InternalError
	case 501:
		return NotImplemented
	case 503:
		return ServiceUnavailable
	case 600:
		return ProtocolError
	case 601:
		return SerializationError
	case 602:
		return RoutingError
	case 603:
		return ConnectionError
	default:
		return InternalError
	}
}

// FromHTTPStatus 从 HTTP 状态码映射到 ErrorCode
func FromHTTPStatus(httpStatus int) ErrorCode {
	switch httpStatus {
	case 400:
		return BadRequest
	case 401:
		return Unauthorized
	case 403:
		return Forbidden
	case 404:
		return NotFound
	case 408:
		return Timeout
	case 500:
		return InternalError
	case 501:
		return NotImplemented
	case 503:
		return ServiceUnavailable
	default:
		if httpStatus >= 400 && httpStatus < 500 {
			return BadRequest
		}
		if httpStatus >= 500 {
			return InternalError
		}
		return InternalError
	}
}

// FromGRPCStatus 从 gRPC 状态码映射到 ErrorCode
func FromGRPCStatus(grpcStatus int) ErrorCode {
	switch grpcStatus {
	case 0: // OK
		return 0
	case 1: // CANCELLED
		return InternalError
	case 2: // UNKNOWN
		return InternalError
	case 3: // INVALID_ARGUMENT
		return BadRequest
	case 4: // DEADLINE_EXCEEDED
		return Timeout
	case 5: // NOT_FOUND
		return NotFound
	case 7: // PERMISSION_DENIED
		return Forbidden
	case 12: // UNIMPLEMENTED
		return NotImplemented
	case 13: // INTERNAL
		return InternalError
	case 14: // UNAVAILABLE
		return ServiceUnavailable
	case 16: // UNAUTHENTICATED
		return Unauthorized
	default:
		return InternalError
	}
}

// FromJSONRPCCode 从 JSON-RPC 错误码映射到 ErrorCode
func FromJSONRPCCode(jsonRpcCode int) ErrorCode {
	switch jsonRpcCode {
	case -32700: // Parse error
		return BadRequest
	case -32600: // Invalid Request
		return BadRequest
	case -32601: // Method not found
		return NotFound
	case -32602: // Invalid params
		return BadRequest
	case -32603: // Internal error
		return InternalError
	default:
		return InternalError
	}
}

// ToHTTPStatus 将 ErrorCode 映射到 HTTP 状态码
func (e ErrorCode) ToHTTPStatus() int {
	switch e {
	case BadRequest:
		return 400
	case Unauthorized:
		return 401
	case Forbidden:
		return 403
	case NotFound:
		return 404
	case Timeout:
		return 408
	case InternalError:
		return 500
	case NotImplemented:
		return 501
	case ServiceUnavailable:
		return 503
	case ProtocolError:
		return 502
	case SerializationError:
		return 400
	case RoutingError:
		return 502
	case ConnectionError:
		return 503
	default:
		return 500
	}
}

// ToJSONRPCCode 将 ErrorCode 映射到 JSON-RPC 错误码
func (e ErrorCode) ToJSONRPCCode() int {
	switch e {
	case BadRequest:
		return -32600
	case NotFound:
		return -32601
	case InternalError, Timeout, ServiceUnavailable:
		return -32603
	case SerializationError:
		return -32700
	default:
		return -32603
	}
}
